src/hotspot/share/gc/shenandoah/shenandoahController.cpp

```diff
@@ -92,9 +97,11 @@ void ShenandoahController::handle_alloc_failure_evac(size_t words) {
   heap->cancel_gc(GCCause::_shenandoah_allocation_failure_evac);
 }
 
-void ShenandoahController::notify_alloc_failure_waiters() {
-  _alloc_failure_gc.unset();
-  _humongous_alloc_failure_gc.unset();
+void ShenandoahController::notify_alloc_failure_waiters(bool clear_alloc_failure) {
+  if (clear_alloc_failure) {
```

Comment:
Why would we not clear the alloc failure? This seems like it would confuse the control thread. Isn't this going to have the control thread attempt to notify alloc failure waiters again when the cycle is finished?